### PR TITLE
[SPARK-13150] [SQL] Fixes flaky Thrift server ADD JAR command test cases

### PR DESCRIPTION
**Pull request by liancheng**
This PR provides a workaround for [SPARK-13157](https://issues.apache.org/jira/browse/SPARK-13157) to fix flaky Thrift server `ADD JAR` commands. The problem is that our Jenkins builds may be executed under a path containing `@` character while the `ADD JAR` command doesn't support this kind of path and simply drops the `@` character. The workaround provided here is to first copy the jar file to be added to a temporary path and then execute `ADD JAR` using this path.

Here we also adds more logs to Thrift server test suites to ease future test failure debugging.

### CODE REVIEW

File: sql/hive-thriftserver/src/test/scala/org/apache/spark/sql/hive/thriftserver/HiveThriftServer2Suites.scala [hunk 1]
```diff
@@ -786,6 +787,7 @@ abstract class HiveThriftServer2Test extends SparkFunSuite with BeforeAndAfterAl
       val builder = new ProcessBuilder(command: _*)
       val captureOutput = (line: String) => diagnosisBuffer.synchronized {
         diagnosisBuffer += line
+        logInfo(line)
```
Comment (liancheng) [inline sql/hive-thriftserver/src/test/scala/org/apache/spark/sql/hive/thriftserver/HiveThriftServer2Suites.scala]: This dumps Thrift server logs to `unit-tests.log`.

File: sql/hive-thriftserver/src/test/scala/org/apache/spark/sql/hive/thriftserver/HiveThriftServer2Suites.scala [hunk 2]
```diff
@@ -493,12 +493,12 @@ class HiveThriftBinaryServerSuite extends HiveThriftJdbcTest {
...
 
-      Seq(
-        s"ADD JAR $jarURL",
+      statement.execute(s"ADD JAR $jarURL")
+      statement.execute(
         s"""CREATE TEMPORARY FUNCTION udtf_count2
            |AS 'org.apache.spark.sql.hive.execution.GenericUDTFCount2'
          """.stripMargin
-      ).foreach(statement.execute)
+      )
```
Comment (liancheng) [inline sql/hive-thriftserver/src/test/scala/org/apache/spark/sql/hive/thriftserver/HiveThriftServer2Suites.scala]: When the original code fails, log lines only say that one of the commands within the `Seq` fails. I'd like to know which one fails.

**Comment by liancheng** [comment]
Just scheduled a build wave, trying to reproduce the test failure.

**Comment by liancheng** [comment]
Found 3 interesting log lines:

```
16/02/02 18:38:00 INFO ParseDriver: Parsing command: ADD JAR file:///home/jenkins/workspace/SparkPullRequestBuilder@2/sql/hive-thriftserver/../hive/src/test/resources/TestUDTF.jar
...
16/02/02 18:38:00 DEBUG HiveClientImpl: Running hiveql 'ADD JAR file:///home/jenkins/workspace/SparkPullRequestBuilder2/sql/hive-thriftserver/../hive/src/test/resources/TestUDTF.jar'
...
16/02/02 18:38:00 ERROR SessionState: file:///home/jenkins/workspace/SparkPullRequestBuilder2/sql/hive-thriftserver/../hive/src/test/resources/TestUDTF.jar does not exist
```

Please notice that the first log line has `SparkPullRequestBuilder@2` in the file path while the 2nd and 3rd ones have `SparkPullRequestBuilder2` Seems that the HiveQL parser somehow drops the `@` character, and then causes execution failure of the `ADD JAR` command.

One possible workaround is to first copy the jar file to a temporary location whose path doesn't have any interesting characters and then execute `ADD JAR` using that path.

cc @davies

**Comment by davies** [comment]
@liancheng This should be a bug, can we reproduce it locally (add @ to the path)?

**Comment by liancheng** [comment]
Actually the `ADD JAR` command doesn't fail, but the jar wasn't successfully added, and the following `CREATE TEMPORARY FUNCTION` command failed.

Verified this behavior using the following Beeline session:

```
> add jar file:///tmp/a@b/TestUDTF.jar;
...
> CREATE TEMPORARY FUNCTION udtf_count2 AS 'org.apache.spark.sql.hive.execution.GenericUDTFCount2';
Error: org.apache.spark.sql.execution.QueryExecutionException: FAILED: Execution Error, return code 1 from org.apache.hadoop.hive.ql.exec.FunctionTask (state=,code=0)
```

The `TestUDTF.jar` is copied from Spark code repo.

**Comment by liancheng** [comment]
My guess is that when a PR Jenkins build is triggered by the Spark PR dashboard and the build is scheduled to any builder other than the 1st builder, which doesn't include the `@n` suffix, this bug is triggered.

**Comment by liancheng** [comment]
Oh, maybe the `@` character is treated as the separator between user/password and host/port in a URL.

**Comment by davies** [comment]
The second line seems weird, where is the `@`?

**Comment by liancheng** [comment]
@davies The `@` character is removed after being parsed by HiveQL parser (`ParserDriver`).

**Comment by liancheng** [comment]
Complete log line snippet:

```
16/02/02 18:38:00 INFO ParseDriver: Parsing command: ADD JAR file:///home/jenkins/workspace/SparkPullRequestBuilder@2/sql/hive-thriftserver/../hive/src/test/resources/TestUDTF.jar
16/02/02 18:38:00 INFO ParseDriver: Parse completed.
16/02/02 18:38:00 DEBUG HiveClientImpl: Running hiveql 'ADD JAR file:///home/jenkins/workspace/SparkPullRequestBuilder2/sql/hive-thriftserver/../hive/src/test/resources/TestUDTF.jar'
16/02/02 18:38:00 DEBUG VariableSubstitution: Substitution is on: JAR file:///home/jenkins/workspace/SparkPullRequestBuilder2/sql/hive-thriftserver/../hive/src/test/resources/TestUDTF.jar
16/02/02 18:38:00 DEBUG Groups: Returning cached groups for 'jenkins'
16/02/02 18:38:00 DEBUG Groups: Returning cached groups for 'jenkins'
16/02/02 18:38:00 DEBUG SessionState: Session is using authorization class class org.apache.hadoop.hive.ql.security.authorization.DefaultHiveAuthorizationProvider
16/02/02 18:38:00 ERROR SessionState: file:///home/jenkins/workspace/SparkPullRequestBuilder2/sql/hive-thriftserver/../hive/src/test/resources/TestUDTF.jar does not exist
```

Full log file can be downloaded [here][1] (`hive-thriftserver/target/unit-tests.log`).

[1]: https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/2506/

**Comment by liancheng** [comment]
I'm going to fix this issue using the workaround mentioned above and file a separate JIRA ticket for the `@` character issue.

**Comment by davies** [comment]
@liancheng Since these tests are ignored, we could wait for the fix (@), do need to rush a work around.

**Comment by liancheng** [comment]
The Thrift server and the `ADD JAR` command are both pretty tricky to get right, I think it would be better to enable these tests first?

**Comment by liancheng** [comment]
Actually I've already pushed the fix to this PR branch...

**Comment by liancheng** [comment]
Filed SPARK-13157 to track the `@` character issue.

**Comment by liancheng** [comment]
Ah, our Jenkins PR builder overrides `java.io.tmpdir` property and the created temporary directory is still under a parent directory like `SparkPullRequestBuilder@2`.

**Comment by liancheng** [comment]
Seems that the `@` removal bug is causing more serious build failures like [this one][1]. My theory is that some test case added a resource whose path containing `@` to `TestHive`. The `ADD` command itself succeeds but the resource isn't successfully added since the `@` character in the path is removed. This breaks internal state of `TestHive` and causes successive test failures.

Thus I agree with @davies that we should have a proper fix of the `@` removal bug rather than adding a workaround. So closing this.

[1]: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/50650/testReport/

**Pull request closed without merging**